compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -412,27 +394,33 @@ fn irrefutable_let_pattern(id: HirId, ls: LetSource, span: Span, tcx: TyCtxt<'_>
         }};
     }
 
+    let is_if_let = || {
+        matches!(
+            tcx.hir().parent_iter(id).nth(1).unwrap().1,
+            hir::Node::Expr(hir::Expr { kind: hir::ExprKind::If { .. }, .. })
+        )
+    };
     tcx.struct_span_lint_hir(IRREFUTABLE_LET_PATTERNS, id, span, |lint| match ls {
-        LetSource::GenericLet => {
-            emit_diag!(lint, "`let`", "`let` is useless", "removing `let`");
-        }
-        LetSource::IfLet => {
+        hir::LetSource::If if is_if_let() => {
             emit_diag!(
                 lint,
                 "`if let`",
                 "`if let` is useless",
                 "replacing the `if let` with a `let`"
             );
         }
-        LetSource::IfLetGuard => {
+        hir::LetSource::If | hir::LetSource::Local => {
```

Comment:
How can removing the `let` make sense for LetSource::Local?

Reply:
I think that is unreachable. I'll add `hir::LetSource::Local => {}`.